text/0000-stabilize-marker-freeze.md

```diff
@@ -0,0 +1,148 @@
+- Feature Name: `stabilize_marker_freeze`
+- Start Date: 2024-05-10
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Stabilize `core::marker::Freeze` in trait bounds.
+
+# Motivation
+[motivation]: #motivation
+
+With 1.78, Rust [stabilized](https://github.com/rust-lang/rust/issues/121250) the requirement that `T: core::marker::Freeze` for `const REF: &'a T = T::const_new();` (a pattern referred to as "static-promotion" regardless of whether `'a = 'static`) to be legal. `T: core::marker::Freeze` indicates that `T` doesn't contain any un-indirected `UnsafeCell`, meaning that `T`'s memory cannot be modified through a shared reference.
+
+The purpose of this change was to ensure that interior mutability cannot affect content that may have been static-promoted in read-only memory, which would be a soundness issue.
+
+However, this new requirement also prevents using static-promotion to allow generics to provide a generic equivalent to `static` (with the distinction that static-promotion doesn't guarantee a unique address for the promoted content). An example of this pattern can be found in `stabby` and `equator`'s shared way of constructing v-tables:
+```rust
+pub trait VTable<'a>: Copy {
+	const VT: &'a Self;
+}
+pub struct VtAccumulator<Tail, Head> {
+	tail: Tail,
+	head: Head,
+}
+impl<Tail: VTable<'a>, Head: VTable<'a>> VTable<'a> for VtAccumulator<Tail, Head> {
+	const VT: &'a Self = &Self {tail: *Tail::VT, head: *Head::VT}; // Doesn't compile since 1.78
+} 
+```
+
+Making `VTable` a subtrait of `core::marker::Freeze` in this example is sufficient to allow this example to compile again, as static-promotion becomes legal again. This is however impossible as of today due to `core::marker::Freeze` being restricted to `nightly`.
+
+Orthogonally to static-promotion, `core::marker::Freeze` can also be used to ensure that transmuting `&T` to a reference to an interior-immutable type (such as `[T; core::mem::size_of::<T>()]`) is sound (as interior-mutation of a `&T` may lead to UB in code using the transmuted reference, as it expects that reference's pointee to never change). This is notably a safety requirement for `zerocopy` and `bytemuck` which currently use their own equivalents of `core::marker::Freeze` to ensure that requirement, which imposes great maintenance pressure on these crates to ensure they support as many types as possible. They would notably benefit from `core::marker::Freeze`'s status as an auto-trait.
```

Comment:
```suggestion
Orthogonally to static-promotion, `core::marker::Freeze` can also be used to ensure that transmuting `&T` to a reference to an interior-immutable type (such as `[u8; core::mem::size_of::<T>()]`) is sound (as interior-mutation of a `&T` may lead to UB in code using the transmuted reference, as it expects that reference's pointee to never change). This is notably a safety requirement for `zerocopy` and `bytemuck` which currently use their own equivalents of `core::marker::Freeze` to ensure that requirement, which imposes great maintenance pressure on these crates to ensure they support as many types as possible. They would notably benefit from `core::marker::Freeze`'s status as an auto-trait.
```

(I assume that's what you meant?)

Also, re: zerocopy and bytemuck (cc @zachs18), I believe it'd be more accurate to say that:
- zerocopy is [currently developing its equivalent](https://docs.rs/zerocopy/0.8.0-alpha.11/zerocopy/trait.Immutable.html)
- bytemuck is [considering how to use `Freeze`](https://github.com/Lokathor/bytemuck/discussions/153)

I'm not sure those distinctions are particularly important, but I figured I'd point it out in case you care about being precise. @zachs18 can comment on whether I've mis-characterized bytemuck's situation.

Finally, zerocopy doesn't need `Freeze` to be an auto-trait. It just needs it to be supported *somehow*. E.g., requiring `impl Freeze for MyType` a la `Copy` would be fine for us.